pkg-r/R/contents_shinychat.R

```diff
@@ -0,0 +1,207 @@
+#' Format ellmer content for shinychat
+#'
+#' @param content An [`ellmer::Content`] object.
+#'
+#' @return Returns text, HTML, or web component tags formatted for use in
+#'   `chat_ui()`.
+#'
+#' @export
+contents_shinychat <- S7::new_generic(
+  "contents_shinychat",
+  "content",
+  function(content) {
+    S7::S7_dispatch()
+  }
+)
+
+S7::method(contents_shinychat, ellmer::Content) <- function(content) {
+  # Fall back to html or markdown
+  html <- ellmer::contents_html(content)
+  if (!is.null(html)) {
+    shiny::HTML(html)
+  } else {
+    ellmer::contents_markdown(content)
+  }
+}
+
+S7::method(contents_shinychat, ellmer::ContentText) <- function(content) {
+  content@text
+}
+
+S7::method(contents_shinychat, ellmer::ContentToolRequest) <- function(
+  content
+) {
+  props <- list(
+    "request-id" = content@id,
+    name = content@name,
+    arguments = jsonlite::toJSON(content@arguments, auto_unbox = TRUE)
+  )
+
+  # Add optional title if present in tool annotations
+  if (!is.null(content@tool@annotations$title)) {
+    props$title <- content@tool@annotations$title
+  }
+
+  # Add optional intent if present in request arguments
+  if (!is.null(content@arguments$.intent)) {
+    props$intent <- content@arguments$.intent
+  }
+
+  # Return structured tag
+  htmltools::tag("shiny-tool-request", props)
+}
+
+S7::method(contents_shinychat, ellmer::ContentToolResult) <- function(content) {
+  # Prepare base props
+  props <- list(
+    request_id = content@request@id,
+    request_call = format(content@request, show = "call"),
+    name = content@request@name,
+    status = if (!is.null(content@error)) "error" else "success",
+    show_request = tolower(!isFALSE(content@extra$display_tool_request))
+  )
+
+  # Add optional title if present
+  if (!is.null(content@request@tool@annotations$title)) {
+    props$title <- content@request@tool@annotations$title
+  }
+
+  # Add optional icon if present
+  icon_deps <- NULL
+  if (!is.null(content@request@tool@annotations$icon)) {
+    props$icon <- content@request@tool@annotations$icon
+    icon_deps <- htmltools::findDependencies(props$icon)
+  }
+
+  # Add optional intent if present
+  if (!is.null(content@request@arguments$.intent)) {
+    props$intent <- content@request@arguments$.intent
+  }
+
+  display_props <- tool_result_display(content)
+  props$value <- display_props$value
+  props$value_type <- display_props$value_type
+
+  names(props) <- gsub("_", "-", names(props))
+
+  htmltools::tag(
+    "shiny-tool-result",
+    compact(list2(!!!props, display_props$deps, icon_deps))
+  )
+}
+
+tool_result_display <- function(content) {
+  # Default fallback
+  res <- list(
+    value = content@value,
+    value_type = "code",
+    deps = NULL
+  )
```

Comment:
If I'm following correctly, this effectively means we do `tag(value = value)` where `value` could be any R object? 

I'm not 100% positive, but I think htmltools will run `as.character()` on `value`, which isn't great for a lot of situations (e.g., data frames).

Maybe this should behave similar to how `renderPrint()` does? Where you effectively get whatever you'd see if you printed the object at the console (i.e., `capture.output(print(value))`)?

Reply:
Fixed in b4aa4fa617396397e06205bd2720ca869a16d27d